frontend/src/framework/components/ParameterListFilter/private-utils/smartNodeSelectorUtils.ts

```diff
@@ -0,0 +1,179 @@
+};
+
+export function fromParameterTypeToNodeName(type: ParameterType): string {
+    if (type === ParameterType.CONTINUOUS) {
+        return ParameterParentNodeNames.CONTINUOUS;
+    }
+    if (type === ParameterType.DISCRETE) {
+        return ParameterParentNodeNames.DISCRETE;
+    }
+    throw new Error(`Parameter type ${type} not supported`);
+}
+
+export function createAndAddNode(treeNodeDataList: TreeDataNode[], nodeName: string, icon?: string): TreeDataNode {
+    const newNode: TreeDataNode = { name: nodeName, description: "", icon: icon };
+    treeNodeDataList.push(newNode);
+    return newNode;
+}
+
+export function findOrCreateNode(treeNodeDataList: TreeDataNode[], nodeName: string, icon?: string): TreeDataNode {
+    const existingNode = treeNodeDataList.find((node) => node.name === nodeName);
+    if (existingNode) {
+        return existingNode;
+    }
+
+    return createAndAddNode(treeNodeDataList, nodeName, icon);
+}
+
+export function addParameterNameAndGroupToTreeDataNodeList(
+    treeNodeDataList: TreeDataNode[],
+    parameter: Parameter,
+    icon?: string
+): void {
+    // Parameter Name
+    const nameParentNode = findOrCreateNode(treeNodeDataList, ParameterParentNodeNames.NAME, icon);
+    if (!nameParentNode.children) {
+        nameParentNode.children = [];
+    }
+    findOrCreateNode(nameParentNode.children, parameter.name);
+
+    // Parameter Group
+    if (parameter.groupName) {
+        const groupParentNode = findOrCreateNode(treeNodeDataList, ParameterParentNodeNames.GROUP, icon);
+        if (!groupParentNode.children) {
+            groupParentNode.children = [];
+        }
+        findOrCreateNode(groupParentNode.children, parameter.groupName);
+    }
+}
+
+export function createTreeDataNodeListFromParameters(
+    parameters: Parameter[],
+    checkIcon?: string,
+    parentIcon?: string
+): TreeDataNode[] {
+    if (parameters.length === 0) {
+        return [];
+    }
+
+    const treeDataNodeList: TreeDataNode[] = [];
+
+    const hasContinuousParameter = parameters.some((parameter) => parameter.type === ParameterType.CONTINUOUS);
+
+    // Node for boolean/state properties on top level
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.CONTINUOUS, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.DISCRETE, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_CONSTANT, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_NONCONSTANT, checkIcon);
+    if (hasContinuousParameter) {
+        createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_LOGARITHMIC, checkIcon);
+        createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_LINEAR, checkIcon);
+    }
+
+    // Add name and group for parameters
+    for (const parameter of parameters) {
+        addParameterNameAndGroupToTreeDataNodeList(treeDataNodeList, parameter, parentIcon);
+    }
```

Comment:
Consider doing a check and adding the parent nodes once (here) instead of checking for every single parameter if the respective parent node (name/group) needs to be added (removes a lot of unnecessary loops). In addition, are the parameter names unique? If so, you could skip the check for existence for the parameter names and just add them instantly while the group names probably need checking. However, you could also first make a set of group names and then add them.

Reply:
I've made a suggestion please review it. Put both parameter name and group name in a set, and then iterate over these to prevent duplicated names/groups. This prevents usage of `findOrCreateNode` method, which is removed.